src/ucs/config/parser.c

```diff
@@ -1066,6 +1066,173 @@ void ucs_config_help_table(char *buf, size_t max, const void *arg)
     snprintf(buf, max, "Table");
 }
 
+/**
+ * Return number of allowed keys in the given config
+ */
+static size_t ucs_config_key_value_count(const ucs_config_key_field_t *cfg_key)
+{
+    size_t count = 0;
+    for (; cfg_key->name; ++cfg_key, ++count);
+    return count;
+}
+
+/**
+ * Given key-value pair, find appropriate destination for storing value in the
+ * temporary values buffer, and save it there
+ */
+static ucs_status_t
+ucs_config_key_value_set(const ucs_config_key_field_t *cfg_key,
+                         const char **values, size_t key_count,
+                         const char *key, const char *value)
+{
+    size_t index;
+
+    if (NULL == value) {
+        /* Empty value means that key is a generic value for all keys */
+        for (index = 0; index < key_count; ++index) {
+            values[index] = key;
+        }
+        return UCS_OK;
+    } else {
+        /* Key configured, save value for specific key */
+        for (index = 0; index < key_count; ++index, ++cfg_key) {
+            if (!strcmp(cfg_key->name, key)) {
+                values[index] = value;
+                return UCS_OK;
+            }
+        }
+        return UCS_ERR_INVALID_PARAM;
+    }
+}
+
+int ucs_config_sscanf_key_value(const char *buf, void *dest, const void *arg)
+{
+    const ucs_config_key_value_param_t *param = arg;
+    size_t buf_len                            = strlen(buf) + 1;
+    const char *TOKEN_DELIM                   = ",";
+    const char *VALUE_DELIM                   = ":";
+    size_t key_count, index;
+    const char **values;
+    char *token, *tokens;
+    const char *key, *value;
+
+    /* values is a temporary array value storage for the keys.
+     * In values array we remember the last applicable setting for each key.
+     * We need it to perform actual setting exactly once per each key */
+    key_count = ucs_config_key_value_count(param->keys);
+    values = ucs_alloca(key_count * sizeof(*values));
+    memset(values, 0, key_count * sizeof(*values));
+
+    /* Writable copy of the input buffer on stack */
+    tokens = ucs_alloca(buf_len);
+    ucs_strncpy_safe(tokens, buf, buf_len);
```

Comment:
maybe use string buffer? it has tokenize api

Reply:
LGTM, applying this change

Reply:
Btw, currently I cannot easily declare string buffer of dynamic size on stack, because `UCS_ARRAY_ALLOC_ONSTACK` explicitly requires a compile-time constant for capacity.
May we can relax this requirement, and instead of checking at compile-time, perform check on runtime by using `ucs_alloca`?
```
 #define UCS_ARRAY_ALLOC_ONSTACK(_array_type, _capacity) \
     ({ \
         typedef ucs_typeof(*((_array_type*)NULL)->buffer) value_t; \
-        UCS_STATIC_ASSERT((_capacity) * sizeof(value_t) <= \
-                          UCS_ALLOCA_MAX_SIZE); \
-        (value_t*)alloca((_capacity) * sizeof(value_t)); \
+        (value_t*)ucs_alloca((_capacity) * sizeof(value_t)); \
     })
```
This would allow us to define arrays of dynamic size on stack
Any concerns?

Reply:
Ok, I tried to use string buffer here, but I don't like the final result.
The reason is that tokenize algorithm of string buffer is less suitable for this particular use case. Basically here we do 2 phase tokenising: first we split on comma delimiter, and then on colon:
```
    tokens = ucs_string_split(tokens, TOKEN_DELIM, 1, &token);
    while (NULL != token) {
        ucs_string_split(token, VALUE_DELIM, 2, &key, &value);
        tokens = ucs_string_split(tokens, TOKEN_DELIM, 1, &token);
    }
```
ucs_string_split is convenient because it does 2 things at once: extracts token AND return string after the current token. This way I can split the current token as I want, and this does not affect behavior.
In string buffer tokenisation is implemented slightly differently: it just returns the current token.
This is fine if we don't split it further.. But if I split it further into key and value, then it breaks the iteration:
```
15:rear:50
token = 15, key = 15, value = NULL
token = rear:50, key = rear, value = 50
token = 50 << This happens because we split on the previous step
```
Well, we can fix it by manually advancing token pointer by certain size, but looks worse than `ucs_string_split`.
So I would keep the existing implementation for simplicity

Reply:
do you think it's possible to extend string_buffer tokenize api to support multi level tokenization? it could be useful.. if the change is too big we can keep the parse it as is

Reply:
Well, look useful to me too. But I would definitely do it in a separate PR, because this one is already above 500 LOC:) I can do it together with array improvements (separate bit field for fixed flag), or in a separate PR.

Btw, what do you think about my previous comment about array on stack: https://github.com/openucx/ucx/pull/9555#discussion_r1432853798
?

Reply:
> Btw, what do you think about my previous comment about array on stack: [#9555 (comment)](https://github.com/openucx/ucx/pull/9555#discussion_r1432853798)

Sounds good to me